File: ios/NativeHttpModule.swift
import Foundation
import Sentry

/**
 * NativeHttpModule (iOS)
 *
 * Makes HTTP requests using URLSession from the iOS native layer.
 *
 * Sentry Cocoa auto-swizzles URLSession when enableNetworkTracking = true
 * (the default), so any URLSession request made while there is an active
 * transaction automatically appears as a child span.
 *
 * Same pattern as Android: we manually start a Sentry transaction before the
 * request so the URLSession span has a parent and shows up in the trace
 * waterfall — even when autoInitializeNativeSdk = false on the JS side.
 */
@objc(NativeHttpModule)
class NativeHttpModule: NSObject {

    @objc
    func makeRequest(_ url: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {

        guard let requestUrl = URL(string: url) else {
            reject("INVALID_URL", "Invalid URL: \(url)", nil)
            return
        }

        // Start a Sentry transaction and bind it to scope so the URLSession
        // swizzling can attach child spans to it.
        let transaction = SentrySDK.startTransaction(
            name: "NativeHttpRequest",
            operation: "http.client.native",
            bindToScope: true
        )
        transaction.setTag(value: "native_module", key: "screen")
        transaction.setData(value: url, key: "requested_url")

        // URLSession auto-instrumented by Sentry Cocoa SDK.
        let task = URLSession.shared.dataTask(with: requestUrl) { data, response, error in
            if let error = error {
                transaction.finish(status: .internalError)
                SentrySDK.capture(error: error)
                reject("HTTP_ERROR", error.localizedDescription, error)
                return
            }

            let body = data.flatMap { String(data: $0, encoding: .utf8) } ?? "empty response"
            transaction.finish(status: .ok)
            resolve(body)
        }
        task.resume()
    }

    @objc
    static func requiresMainQueueSetup() -> Bool { false }
}

File: scripts/NativeHttpModule.kt
package com.sentryrntest

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import io.sentry.ITransaction
import io.sentry.Sentry
import io.sentry.SpanStatus
import io.sentry.TransactionOptions
import io.sentry.android.okhttp.SentryOkHttpEventListener
import io.sentry.android.okhttp.SentryOkHttpInterceptor
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * NativeHttpModule
 *
 * Makes HTTP requests using OkHttp from the Android native layer.
 *
 * SentryOkHttpInterceptor  — adds the request as a child span of the active
 *                            Sentry transaction, and propagates trace headers.
 * SentryOkHttpEventListener — adds detailed timing child spans (DNS, connect,
 *                             TLS, send headers, response body, etc.).
 *
 * Because autoInitializeNativeSdk = false in JS Sentry.init(), the JS
 * navigation transaction does NOT become the active transaction in the native
 * layer.  To ensure the OkHttp spans have a parent (and therefore appear in
 * a trace waterfall), we manually start a Sentry transaction here, bind it
 * to scope, make the request, then finish the transaction.
 *
 * Result in Sentry:
 *   Transaction: "NativeHttpRequest"  (op: "http.client.native")
 *     └─ Span: GET https://...         (from SentryOkHttpInterceptor)
 *          ├─ dns                       (from SentryOkHttpEventListener)
 *          ├─ connect
 *          ├─ tls_handshake
 *          ├─ send_request_headers
 *          └─ response_body_read
 */
class NativeHttpModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    // OkHttpClient is shared across calls — creating it once is the OkHttp recommendation.
    // Both Sentry interceptors are registered here.
    private val okHttpClient: OkHttpClient = OkHttpClient.Builder()
        .addInterceptor(SentryOkHttpInterceptor())          // creates HTTP span + trace headers
        .eventListener(SentryOkHttpEventListener())         // adds detailed timing sub-spans
        .build()

    override fun getName(): String = "NativeHttp"

    @ReactMethod
    fun makeRequest(url: String, promise: Promise) {
        // Start a Sentry transaction in the native layer.
        // isBindToScope = true → makes this the active transaction so
        // SentryOkHttpInterceptor can attach child spans to it.
        val txOptions = TransactionOptions().apply { isBindToScope = true }
        val transaction: ITransaction = Sentry.startTransaction(
            "NativeHttpRequest",   // name shown in Sentry Performance UI
            "http.client.native",  // op — use http.client convention
            txOptions
        )
        transaction.setTag("screen", "native_module")
        transaction.setData("requested_url", url)

        // Run the network call off the main thread.
        Thread {
            try {
                val request = Request.Builder()
                    .url(url)
                    .build()

                okHttpClient.newCall(request).execute().use { response ->
                    val body = response.body?.string() ?: "empty response"
                    transaction.finish(SpanStatus.OK)
                    promise.resolve(body)
                }
            } catch (e: Exception) {
                transaction.finish(SpanStatus.INTERNAL_ERROR)
                Sentry.captureException(e)
                promise.reject("HTTP_ERROR", e.message ?: "Unknown error", e)
            }
        }.start()
    }
}
